src/core/value/constant.jl

```diff
@@ -102,53 +102,175 @@ export ConstantAggregateZero
+    ConstantArray(ConstantFP.(data, Ref(ctx)), HalfType(ctx))
+ConstantArray(data::AbstractArray{Float32,N}, ctx::Context=GlobalContext()) where {N} =
+    ConstantArray(ConstantFP.(data, Ref(ctx)), FloatType(ctx))
+ConstantArray(data::AbstractArray{Float64,N}, ctx::Context=GlobalContext()) where {N} =
+    ConstantArray(ConstantFP.(data, Ref(ctx)), DoubleType(ctx))
+
+# convert back to known array types
+function Base.collect(ca::ConstantArray)
+    constants = Array{Value}(undef, size(ca))
+    for I in CartesianIndices(size(ca))
+        @inbounds constants[I] = ca[Tuple(I)...]
+    end
+    return constants
+end
+
+# array interface
 Base.eltype(ca::ConstantArray) = eltype(llvmtype(ca))
-Base.convert(::Type{Array{T,1}}, ca::ConstantArray) where {T<:Integer} =
-    [convert(T,ConstantInt(ca[i])) for i in 1:length(ca)]
-Base.convert(::Type{Array{T,1}}, ca::ConstantArray) where {T<:AbstractFloat} =
-    [convert(T,ConstantFP(ca[i])) for i in 1:length(ca)]
+function Base.size(ca::ConstantArray)
+    dims = Int[]
+    typ = llvmtype(ca)
+    while typ isa ArrayType
+        push!(dims, length(typ))
+        typ = eltype(typ)
+    end
+    return Tuple(dims)
+end
+Base.length(ca::ConstantArray) = prod(size(ca))
+Base.axes(ca::ConstantArray) = Base.OneTo.(size(ca))
+
+function Base.getindex(ca::ConstantArray, idx::Integer...)
+    # multidimensional arrays are represented by arrays of arrays,
+    # which we need to 'peel back' by looking at the operand sets.
+    # for the final dimension, we use LLVMGetElementAsConstant
+    @boundscheck Base.checkbounds_indices(Base.Bool, axes(ca), idx) ||
+        throw(BoundsError(ca, idx))
+    I = CartesianIndices(size(ca))[idx...]
+    for i in Tuple(I)
+        if isempty(operands(ca))
+            ca = LLVM.Value(API.LLVMGetElementAsConstant(ca, i-1))
+        else
+            ca = (Base.@_propagate_inbounds_meta; operands(ca)[i])
+        end
+    end
+    return ca
+end
+
+# structs
 
 @checked struct ConstantStruct <: ConstantAggregate
     ref::API.LLVMValueRef
 end
 identify(::Type{Value}, ::Val{API.LLVMConstantStructValueKind}) = ConstantStruct
 
+ConstantStructOrAggregateZero(value) = Value(value)::Union{ConstantStruct,ConstantAggregateZero}
+
+# anonymous
+ConstantStruct(values::Vector{<:Constant}; packed::Core.Bool=false) =
+    ConstantStructOrAggregateZero(API.LLVMConstStruct(values, length(values), convert(Bool, packed)))
+ConstantStruct(values::Vector{<:Constant}, ctx::Context; packed::Core.Bool=false) =
+    ConstantStructOrAggregateZero(API.LLVMConstStructInContext(ctx, values, length(values), convert(Bool, packed)))
+
+# named
+ConstantStruct(typ::StructType, values::Vector{<:Constant}) =
+    ConstantStructOrAggregateZero(API.LLVMConstNamedStruct(typ, values, length(values)))
+
+# create a ConstantStruct from a Julia object
+function ConstantStruct(value::T, ctx::Context=GlobalContext(); name=String(nameof(T)),
```

Comment:
My use case involves creating `ConstantArray` of `ConstantStruct` to initialize a `GlobalVariable`. For this to work I need the types of `ConstantStruct` to match those of the `GlobalVariable`. To implement this without resorting to the lower level API, I'd like to have an optional `struct_typ::Union{StructType,Nothing}=nothing` argument.

Reply:
Just use the lower level API for now. I'm not happy with the design, and I'd rather not keep on piling on constructors here only to deprecate them later (requiring a breaking release). At the same time, the lower-level API is unlikely to change.